lib/services/authority.go

```diff
@@ -50,6 +70,107 @@ func NewCertAuthority(
 	})
 }
 
+// ValidateCertAuthority validates the CertAuthority
+func ValidateCertAuthority(ca CertAuthority) (err error) {
+	if err = ca.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+	switch ca.GetType() {
+	case UserCA, HostCA:
+		err = checkUserOrHostCA(ca)
+	case types.JWTSigner:
+		err = checkJWTKeys(ca)
+	}
+	return trace.Wrap(err)
+}
+
+func checkUserOrHostCA(ca CertAuthority) error {
+	if len(ca.GetCheckingKeys()) == 0 {
+		return trace.BadParameter("certificate authority missing SSH public keys")
+	}
+	if len(ca.GetTLSKeyPairs()) == 0 {
+		return trace.BadParameter("certificate authority missing TLS key pairs")
+	}
+	if _, err := ca.Checkers(); err != nil {
+		return trace.Wrap(err)
+	}
+	if _, err := ca.Signers(); err != nil {
+		return trace.Wrap(err)
+	}
+	// This is to force users to migrate
+	if len(ca.GetRoles()) != 0 && len(ca.GetRoleMap()) != 0 {
+		return trace.BadParameter("should set either 'roles' or 'role_map', not both")
+	}
+	err := ca.GetRoleMap().Check()
+	return trace.Wrap(err)
+}
+
+func checkJWTKeys(ca CertAuthority) error {
```

Comment:
I think we can keep this check in the client library code:
- `jwt.New` validation is basically just checking that `ca.ClusterName` is set and `privateKey` is not nil (it's not nil if `utils.ParsePrivateKey` succeeds); you could easily do those checks here
- `utils.ParsePublicKey/ParsePrivateKey` only depend on stdlib and can be moved to a package under `api/` without adding dependencies (this is similar to what the AWS SDK does - implement a bunch of helper libs in the module instead of as external deps)

similarly, `checkUserOrHostCA` doesn't depend on anything external and can be kept in `api/`

Reply:
@awly If that logic is moved out to the `api/`, it will have to be implemented in other languages as well.

Reply:
Good point.
My only concern is: it should be easy for a new junior engineer to figure out where to find existing validation logic and what's the appropriate place to add new validation.

Reply:
@awly I completely agree that `services` and `types` need more refactoring. I've been thinking through some solutions. My favorite solution right now is to extend the interfaces in `api/types` to `lib/services/types`, which will have more business logic methods defined on them.

For example, `CertAuthority` would be extended as `CertAuthorityService` in `lib/services`. And we can use this interface across teleport, and leave `api/types` specifically to hold our gRPC defined types for client side usage and server side unmarshalling.
```
type CertAuthorityService interface {
    types.CertAuthority
    Validate() error
    TLSCerts ... 
    JWTSigner ....
}
```

I think that this belongs in a RFD for more discussion, but right now I'm focused on getting the client to a stage where it can be used by customers. But trust me, the API is my baby now, I plan on dressing it up as soon as I have time.

And to your point about it confusing junior engineers, this is already the case, because we call validation logic in seemingly random places. I definitely want to make the calls consistent like you mentioned in this comment https://github.com/gravitational/teleport/pull/5261#pullrequestreview-571541754